Adding better error logging when roles block host user creation

eriktate opened the pull request:
Related to #40370
Fixes #46252

Adds more descriptive error logging when a role blocks user creation on a host, including the name of the role that blocked them. The role included in the message isn't necessarily the only role blocking access since the `AccessChecker` fails fast, but this should provide enough info for debugging purposes.

changelog: Fixed an issue preventing session joining while host user creation was in use.

eriktate committed:
adding which role caused auth failure to the resulting error when checking for create_host_user_mode and surfacing for helpful error logging server side

eriktate commented on lib/services/access_checker.go:
```diff
@@ -1008,7 +1008,7 @@ func (a *accessChecker) HostUsers(s types.Server) (*HostUsersInfo, error) {
 		// user, the user should not be allowed on
 		if createHostUserMode == types.CreateHostUserMode_HOST_USER_MODE_OFF {
-			return nil, trace.AccessDenied("user is not allowed to create host users")
+			return nil, trace.AccessDenied("role %q prevents creating host users", role.GetName())
```
Is there any concern around exposing the role that blocked an action like this? It was requested in the original issue and it definitely makes troubleshooting a lot simpler, but I'm not 100% sure if it's wise to write the role back to the client? Nothing bad immediately comes to mind, but just calling it out in case someone else can think of a good reason not to

nklaassen commented on lib/srv/sess.go:
```diff
@@ -282,7 +282,7 @@ func (s *SessionRegistry) TryCreateHostUser(identityContext IdentityContext) (cr
 		if trace.IsAccessDenied(err) {
 			log.Warnf("Unable to create host users: %v", err)
-			return false, nil, nil
+			return false, nil, trace.Wrap(err)
```
Unfortunately you can't just return the error here,`TryCreateHostUser` is called for every incoming SSH connection, if the user isn't allowed to create host users but the user already exists, that's fine and we should allow the connection. I'm pretty sure that's why some of the tests are failing on this PR.

We only really want to return the error if the host user doesn't already exist. I think the existing code "tries" to return this error below with "Insufficient permission for host user creation", but there is a logic bug, `trace.IsAccessDenied(err)` will always be false or else the function would have already returned

eriktate replied:
Great call out :+1: I should've noticed we were specifically allowing connections a couple of lines down  🤦 I think these checks are _slightly_ less confusing now, and I also fixed up an existing case where a failed check against the access checker would result in an error being returned regardless of whether or not the host user already exists

rosstimothy reviewed (commented):
Can we add test coverage?

nklaassen commented on lib/srv/sess.go:
```diff
@@ -278,19 +278,17 @@ func (s *SessionRegistry) TryCreateHostUser(identityContext IdentityContext) (cr
-			log.Warnf("Unable to create host users: %v", err)
-			return false, nil, nil
+	ui, accessErr := identityContext.AccessChecker.HostUsers(s.Srv.GetInfo())
+	existsErr := s.users.UserExists(identityContext.Login)
+	if existsErr != nil {
+		if trace.IsAccessDenied(accessErr) {
+			return false, nil, trace.WrapWithMessage(accessErr, "Insufficent permission for host user creation")
```
We need to be really careful not to introduce a regression here. Previously, we returned no error if `trace.IsAccessDenied(accessErr) && existsErr != nil`, now we do. Which I know is the entire point of this PR but I think there are cases where the SSH connection should actually succeed:
- the `UserExists` implementation might return an error even though the user does, in fact, exist (it would on non-linux, but `GetCreateHostUser()` above would protect us there. Maybe there are other cases)
- the user doesn't exist now, but will be created just-in-time by a PAM module https://goteleport.com/docs/enroll-resources/server-access/guides/ssh-pam/#create-local-unix-users-on-login

I think we should probably only return the access denied error if the SSH connection actually fails because the user doesn't exist. Hopefully there's a way to catch that error and add the extra context

eriktate commented on lib/srv/regular/sshserver.go:
```diff
@@ -1264,20 +1264,24 @@ func (s *Server) HandleNewConn(ctx context.Context, ccx *sshutils.ConnectionCont
 
 	// Create host user.
-	created, userCloser, err := s.termHandlers.SessionRegistry.TryCreateHostUser(identityContext)
+	created, userCloser, err := s.termHandlers.SessionRegistry.UpsertHostUser(identityContext)
```
Dropping the `Try` nomenclature since these functions return errors. Eventually I'd like to change the public interface of `usermgmt` to avoid this confusion, but for now renaming it is easy

nklaassen commented on lib/srv/regular/sshserver.go:
```diff
@@ -1264,20 +1264,22 @@ func (s *Server) HandleNewConn(ctx context.Context, ccx *sshutils.ConnectionCont
+	created, userCloser, err := s.termHandlers.SessionRegistry.UpsertHostUser(identityContext)
 	if err != nil {
-		return ctx, trace.Wrap(err)
+		log.Warnf("error while creating host users: %s", err)
```
I'm not sure what this PR is really getting us now, weren't we already logging warnings? But the user rarely sees those logs. If we are going to close the linked issue I would love to see in the PR description what the user actually sees when encountering the error now and how we've improved it

I'm also against logging at the warning level for conditions that we totally expect, like not creating a host user for a teleport user that has no permissions to create host users, which is the expected path for most SSH connections probably

I wonder if we could instead update `(*Server).replyError` to try to detect errors related to the user not existing and add context to the error we send back to the user there

nklaassen replied:
Oh I just saw we're not actually trying to close https://github.com/gravitational/teleport/issues/40370 with this anymore

eriktate replied:
Yeah part of this solution ended up overlapping with a bugfix for the second issue now linked in the description. I plan on opening another PR with the work to propagate errors appropriately to the client. For now I can at least reduce the log level, because I agree we shouldn't be warning for something expected out of a healthy system  :+1:

eriktate commented on lib/srv/sess.go:
```diff
@@ -265,32 +269,42 @@ func (s *SessionRegistry) TryWriteSudoersFile(identityContext IdentityContext) (
+		return false, nil, trace.Wrap(accessErr)
 	}
 
 	userCloser, err := s.users.UpsertUser(identityContext.Login, *ui)
```
I think this is effectively returning the error only for unexpected errors, but it's written in a way that's confusing. I tried reworking this a little bit in order to make it more clear what errors we're checking for and whether or not we return unexpected errors

nklaassen replied:
> I think this is effectively returning the error only for unexpected errors

If this is true then I believe the caller should return the error rather than just logging it. We should only ignore/log/not return errors that we expect

But currently I believe it's not quite true, because above we return an AccessDenied error when the user doesn't exist, but that could be totally expected if the user is going to be created by a PAM module

eriktate commented on lib/srv/sess_test.go:
```diff
@@ -1428,4 +1429,301 @@ func mockSSHSession(t *testing.T) *tracessh.Session {
+type fakeHostUsersBackend struct {
+	HostUsers
+
+	calls        map[string]int
```
I noticed this in some other tests but I didn't realize this was an "official" convention. I've updated these tests and I'll keep that in mind for future ones as well, thanks for the tip!

eriktate commented:
@nklaassen I think I've addressed all of your comments. Let me know if there's anything else you'd like to see changed, otherwise I'll plan on merging this tomorrow afternoon :+1:

nklaassen commented on lib/srv/sess.go:
```diff
@@ -265,42 +269,51 @@ func (s *SessionRegistry) TryWriteSudoersFile(identityContext IdentityContext) (
+		existsErr := s.users.UserExists(identityContext.Login)
+		if existsErr != nil {
+			if trace.IsNotFound(existsErr) {
+				return false, nil, trace.WrapWithMessage(accessErr, "insufficient permissions for host user creation")
```
In this case the SSH connection is likely to fail, but it won't if the user is created JiT by a PAM module
```suggestion
				s.log.Warnf("host user %q does not exist and Teleport user has insufficient permissions for host user creation: %v", identityContext.Login, err)
				return false, nil, nil
```

eriktate replied:
In the context of this function I think this is still an error. Since we don't have a way of knowing whether or not this failure is okay, we should let the caller ultimately decide what to do about it. I agree though that there will be plenty of valid cases where this happens, and I think that gets into the follow up PR to add error context we can report on once we know if the connection ultimately fails

nklaassen commented on lib/srv/regular/sshserver.go:
```diff
@@ -1264,20 +1264,22 @@ func (s *Server) HandleNewConn(ctx context.Context, ccx *sshutils.ConnectionCont
+	created, userCloser, err := s.termHandlers.SessionRegistry.UpsertHostUser(identityContext)
 	if err != nil {
-		return ctx, trace.Wrap(err)
+		log.Infof("error while creating host users: %s", err)
```
I suggest making the function only return unexpected errors and returning them here
```suggestion
		return ctx, trace.Wrap(err, "creating host user")
```

nklaassen replied:
I'm operating under the assumption that nobody is ever going to see these logs, especially if they don't have access to the box. But I agree that we should try to allow access if we can. Hopefully we can add some visibility in a later PR if the SSH connection does fail

nklaassen commented on lib/srv/regular/sshserver.go:
```diff
@@ -1264,20 +1264,22 @@ func (s *Server) HandleNewConn(ctx context.Context, ccx *sshutils.ConnectionCont
+	sudoersCloser, err := s.termHandlers.SessionRegistry.WriteSudoersFile(identityContext)
 	if err != nil {
-		return ctx, trace.Wrap(err)
+		log.Warnf("error while writing sudoers: %s", err)
```
the function only returns unexpected errors, I suggest returning them
```suggestion
		return ctx, trace.Wrap(err, "writing sudoers")
```

eriktate replied:
Tim and I talked about this some last week and decided that we probably shouldn't fail the connection for any error related to host user creation or writing sudoers. Even unexpected ones. We could return the error here but I think we'd just end up ignoring it up the chain where it would be harder to determine which errors are safe to ignore

nklaassen commented on lib/srv/sess.go:
```diff
@@ -265,42 +269,51 @@ func (s *SessionRegistry) TryWriteSudoersFile(identityContext IdentityContext) (
-	existsErr := s.users.UserExists(identityContext.Login)
-	if trace.IsAccessDenied(err) && existsErr != nil {
-		return false, nil, trace.WrapWithMessage(err, "Insufficient permission for host user creation")
+	if accessErr != nil {
+		return false, nil, trace.Wrap(accessErr)
```
IIUC you could get into this branch if `trace.IsAccessDenied(accessErr) && existsErr == nil`, which just means the user already exists and we're not allowed to create it, which is totally expected most of the time, so I wouldn't return an error in that case

probably need to make this `if trace.IsAccessDenied(accessErr) { /*...*/ } elseif err != nil { return false, nil, trace.Wrap(accessErr) }`

would be great to have test coverage for this

nklaassen replied:
oh we do have test coverage, nice! i just think it should assert that it returns a nil error error

eriktate replied:
I think this would actually be if `!trace.IsAccessDenied(accessErr)`, so we're just reporting an unexpected error while trying to check access

nklaassen commented on lib/srv/sess_test.go:
```diff
@@ -1428,4 +1429,288 @@ func mockSSHSession(t *testing.T) *tracessh.Session {
+			},
+
+			expectCreated: false,
+			expectErrIs:   trace.AccessDenied("test"),
```
I suggest this is not an error and it should return nil
```suggestion
```

eriktate replied:
[Similar to this](https://github.com/gravitational/teleport/pull/46091/files/61e8c3cd77375627017c759fe72f28efaa65b965#r1754498745), I'm not sure we should assume these aren't errors in this context. If we could verify that they weren't at this level I would agree, but since we only know for sure at a later time I think this case (and the one below) should still return errors

nklaassen replied:
I guess I was a bit hung up on this because it's, like, the default way that teleport works. for probably 99% of successful SSH connections the host user already exists and the teleport user is not allowed to create host users. I wish we were differentiating "this is definitely an error" from "this is 99% probably fine but if the connection fails for a specific reason this might be why". But for now I guess it's fine to keep returning it as an error that only gets logged

nklaassen commented on lib/srv/sess_test.go:
```diff
@@ -1428,4 +1429,288 @@ func mockSSHSession(t *testing.T) *tracessh.Session {
+
+			expectCreated:     false,
+			expectUsers:       make(map[string][]string),
+			expectErrIs:       trace.AccessDenied("test"),
+			expectErrContains: "insufficient permissions for host user creation",
```
I suggest this is not an error and it should return nil. But it should log and maybe you can test for that
```suggestion
```

nklaassen reviewed (approved):
alright thanks for putting up with all my comments. I see the net improvement here, let's ship it, hopefully we can add more visibility later on if the connection actually fails
